File: download_sentinel2_gee.js
/*
 * Download Cloud-Free Sentinel-2 Imagery - Google Earth Engine Code Editor
 *
 * This script downloads cloud-free Sentinel-2 RGB composites for multiple locations.
 *
 * HOW TO USE:
 * 1. Open Google Earth Engine Code Editor: https://code.earthengine.google.com/
 * 2. Copy-paste this entire script
 * 3. Modify the LOCATIONS configuration below
 * 4. Click "Run"
 * 5. Check the "Tasks" tab to start exports
 *
 * FEATURES:
 * - Automatic cloud filtering
 * - Median composite (removes clouds)
 * - Quality assessment
 * - Multiple locations support
 * - Visual preview on map
 */

// ============================================================================
// CONFIGURATION - MODIFY THIS SECTION
// ============================================================================

var CONFIG = {
  // Date range for imagery
  START_DATE: '2023-06-01',
  END_DATE: '2023-09-30',

  // Cloud filtering
  MAX_CLOUD_PERCENTAGE: 10,  // Maximum cloud percentage (0-100)

  // Export settings
  EXPORT_SCALE: 10,  // Resolution in meters (10m for Sentinel-2)
  EXPORT_CRS: 'EPSG:3857',  // Coordinate system (Web Mercator)
  EXPORT_FOLDER: 'RSVQA_Sentinel2',  // Google Drive folder

  // Visualization parameters
  VIS_MIN: 0,
  VIS_MAX: 3000,
  VIS_GAMMA: 1.4
};

// Define your locations here
// Format: [lon_min, lat_min, lon_max, lat_max]
var LOCATIONS = [
  {
    name: 'newyork_manhattan',
    bbox: [-74.05, 40.7, -73.95, 40.8],
    description: 'New York City - Manhattan'
  },
  {
    name: 'iowa_farmland',
    bbox: [-93.65, 41.55, -93.55, 41.65],
    description: 'Iowa agricultural area'
  },
  {
    name: 'amazon_rainforest',
    bbox: [-60.1, -3.1, -60.0, -3.0],
    description: 'Amazon rainforest',
    maxCloud: 15  // Higher cloud threshold for tropical areas
  },
  {
    name: 'california_coast',
    bbox: [-122.5, 37.7, -122.4, 37.8],
    description: 'San Francisco Bay Area'
  },
  {
    name: 'arizona_desert',
    bbox: [-111.1, 33.4, -111.0, 33.5],
    description: 'Arizona desert',
    maxCloud: 5  // Lower cloud threshold for desert
  }
];

// ============================================================================
// FUNCTIONS
// ============================================================================

/**
 * Get cloud-free Sentinel-2 composite
 */
function getCloudFreeComposite(roi, startDate, endDate, maxCloudPct) {
  // Load Sentinel-2 Surface Reflectance collection
  var collection = ee.ImageCollection('COPERNICUS/S2_SR')
    .filterBounds(roi)
    .filterDate(startDate, endDate)
    .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', maxCloudPct))
    .select(['B4', 'B3', 'B2']);  // RGB bands

  // Count available images
  var count = collection.size();
  print('  Images found:', count);

  // Create median composite (removes clouds)
  var composite = collection.median().clip(roi);

  return composite;
}

/**
 * Assess image quality
 */
function assessQuality(image, roi, locationName) {
  var stats = image.reduceRegion({
    reducer: ee.Reducer.mean().combine({
      reducer2: ee.Reducer.stdDev(),
      sharedInputs: true
    }),
    geometry: roi,
    scale: 30,
    maxPixels: 1e9
  });

  print('  Quality Assessment for ' + locationName + ':');
  print('    Mean (B4):', stats.get('B4_mean'));
  print('    StdDev (B4):', stats.get('B4_stdDev'));

  // Quality checks
  var mean = ee.Number(stats.get('B4_mean')).getInfo();
  var std = ee.Number(stats.get('B4_stdDev')).getInfo();

  if (mean < 100) {
    print('    ⚠ Warning: Dark image (mean < 100)');
  } else if (mean > 2500) {
    print('    ⚠ Warning: Bright image (mean > 2500)');
  } else {
    print('    ✓ Good brightness');
  }

  if (std < 50) {
    print('    ⚠ Warning: Low contrast (std < 50)');
  } else {
    print('    ✓ Good contrast');
  }
}

/**
 * Visualize image on map
 */
function visualizeOnMap(image, roi, locationName) {
  var visParams = {
    min: CONFIG.VIS_MIN,
    max: CONFIG.VIS_MAX,
    gamma: CONFIG.VIS_GAMMA
  };

  Map.addLayer(image, visParams, locationName);
  Map.addLayer(roi, {color: 'red'}, locationName + ' - ROI', false);
}

/**
 * Export image to Google Drive
 */
function exportToDrive(image, roi, locationName) {
  // Visualize for export (scale to 0-255)
  var exportImage = image.visualize({
    min: CONFIG.VIS_MIN,
    max: CONFIG.VIS_MAX,
    gamma: CONFIG.VIS_GAMMA
  });

  Export.image.toDrive({
    image: exportImage,
    description: locationName + '_sentinel2',
    folder: CONFIG.EXPORT_FOLDER,
    scale: CONFIG.EXPORT_SCALE,
    region: roi,
    crs: CONFIG.EXPORT_CRS,
    fileFormat: 'GeoTIFF',
    maxPixels: 1e9,
    formatOptions: {
      cloudOptimized: true
    }
  });

  print('  ✓ Export task created: ' + locationName);
}

/**
 * Process a single location
 */
function processLocation(location) {
  print('\n' + '='.repeat(70));
  print('Processing: ' + location.name);
  print('Description: ' + location.description);
  print('='.repeat(70));

  // Create region of interest
  var roi = ee.Geometry.Rectangle(location.bbox);
  print('Bounding box:', location.bbox);

  // Get cloud threshold (use custom or default)
  var maxCloudPct = location.maxCloud || CONFIG.MAX_CLOUD_PERCENTAGE;
  print('Max cloud percentage:', maxCloudPct + '%');

  // Get composite
  var composite = getCloudFreeComposite(
    roi,
    CONFIG.START_DATE,
    CONFIG.END_DATE,
    maxCloudPct
  );

  // Assess quality
  assessQuality(composite, roi, location.name);

  // Visualize on map
  visualizeOnMap(composite, roi, location.name);

  // Create export task
  exportToDrive(composite, roi, location.name);
}

// ============================================================================
// MAIN EXECUTION
// ============================================================================

print('╔═══════════════════════════════════════════════════════════════════╗');
print('║  Sentinel-2 Cloud-Free Image Downloader                          ║');
print('║  For RSVQA Custom Dataset                                        ║');
print('╚═══════════════════════════════════════════════════════════════════╝');

print('\nConfiguration:');
print('  Date range:', CONFIG.START_DATE, 'to', CONFIG.END_DATE);
print('  Default max cloud:', CONFIG.MAX_CLOUD_PERCENTAGE + '%');
print('  Export scale:', CONFIG.EXPORT_SCALE + 'm');
print('  Export folder:', CONFIG.EXPORT_FOLDER);

print('\nLocations to process:', LOCATIONS.length);

// Clear map
Map.clear();

// Process each location
LOCATIONS.forEach(function(location) {
  processLocation(location);
});

// Center map on first location
var firstROI = ee.Geometry.Rectangle(LOCATIONS[0].bbox);
Map.centerObject(firstROI, 11);

// Print instructions
print('\n' + '='.repeat(70));
print('NEXT STEPS:');
print('='.repeat(70));
print('1. Check the preview on the map (toggle layers on the right)');
print('2. Go to the "Tasks" tab (top-right)');
print('3. Click "RUN" for each export task');
print('4. Wait for tasks to complete (5-30 minutes)');
print('5. Download from Google Drive folder: ' + CONFIG.EXPORT_FOLDER);
print('6. Run: python prepare_sentinel2_data.py --input_dir ./downloads --output_dir ./dataset');
print('='.repeat(70));

// ============================================================================
// OPTIONAL: ADDITIONAL HELPER FUNCTIONS
// ============================================================================

/**
 * Calculate optimal date range for a location (find least cloudy period)
 * Uncomment and run this function separately to find best dates
 */
/*
function findBestDateRange(location, year) {
  var roi = ee.Geometry.Rectangle(location.bbox);

  // Check each month
  for (var month = 1; month <= 12; month++) {
    var startDate = year + '-' + String(month).padStart(2, '0') + '-01';
    var endDate = year + '-' + String(month).padStart(2, '0') + '-28';

    var collection = ee.ImageCollection('COPERNICUS/S2_SR')
      .filterBounds(roi)
      .filterDate(startDate, endDate);

    var count = collection.size();
    var avgCloud = collection.aggregate_mean('CLOUDY_PIXEL_PERCENTAGE');

    print('Month ' + month + ':',
          'Images:', count.getInfo(),
          'Avg Cloud:', avgCloud.getInfo().toFixed(1) + '%');
  }
}

// Example usage:
// findBestDateRange(LOCATIONS[0], '2023');
*/

/**
 * Visualize cloud coverage distribution
 */
/*
function visualizeCloudDistribution(location) {
  var roi = ee.Geometry.Rectangle(location.bbox);

  var collection = ee.ImageCollection('COPERNICUS/S2_SR')
    .filterBounds(roi)
    .filterDate(CONFIG.START_DATE, CONFIG.END_DATE)
    .select('CLOUDY_PIXEL_PERCENTAGE');

  var chart = ui.Chart.feature.histogram({
    features: collection,
    property: 'CLOUDY_PIXEL_PERCENTAGE',
    maxBuckets: 20
  }).setOptions({
    title: 'Cloud Coverage Distribution - ' + location.name,
    hAxis: {title: 'Cloud Percentage'},
    vAxis: {title: 'Number of Images'}
  });

  print(chart);
}

// Example usage:
// visualizeCloudDistribution(LOCATIONS[0]);
*/

// ============================================================================
// TIPS FOR SELECTING GOOD LOCATIONS
// ============================================================================

/*
TIPS:
1. Use Google Earth to preview areas first
2. Choose areas with interesting features (buildings, water, vegetation mix)
3. Start with 0.1° × 0.1° bounding boxes (~11km × 11km)
4. For tropical/coastal areas: increase maxCloud to 15-20%
5. For desert/arid areas: decrease maxCloud to 5%
6. Summer months (Jun-Sep in Northern Hemisphere) have less cloud cover
7. Avoid areas at high latitudes in winter (very dark)

EXAMPLE INTERESTING LOCATIONS:

Urban:
- New York: [-74.05, 40.7, -73.95, 40.8]
- Los Angeles: [-118.3, 34.0, -118.2, 34.1]
- Chicago: [-87.7, 41.8, -87.6, 41.9]

Agricultural:
- Iowa farmland: [-93.65, 41.55, -93.55, 41.65]
- Central Valley CA: [-120.5, 36.8, -120.4, 36.9]

Natural:
- Grand Canyon: [-112.2, 36.0, -112.1, 36.1]
- Yellowstone: [-110.8, 44.5, -110.7, 44.6]

Coastal:
- San Francisco Bay: [-122.5, 37.7, -122.4, 37.8]
- Florida Keys: [-81.5, 24.6, -81.4, 24.7]

Mixed:
- Phoenix AZ: [-112.1, 33.4, -112.0, 33.5]
- Seattle WA: [-122.4, 47.5, -122.3, 47.6]
*/

File: download_sentinel2_gee.py
#!/usr/bin/env python3
"""
Download Cloud-Free Sentinel-2 Imagery from Google Earth Engine

This script downloads cloud-free Sentinel-2 RGB composites for multiple
geographic locations. Features include:
- Automatic cloud filtering
- Median composite creation (removes clouds)
- Multiple location support
- Quality assessment
- Direct export to Google Drive or local download

Requirements:
    pip install earthengine-api geemap

Setup:
    1. Create GEE account: https://earthengine.google.com/signup/
    2. Authenticate: earthengine authenticate
    3. Run this script

Usage:
    python download_sentinel2_gee.py --config locations.json
    python download_sentinel2_gee.py --interactive
"""

import ee
import geemap
import argparse
import json
from datetime import datetime, timedelta
from pathlib import Path
import time


class Sentinel2Downloader:
    """Download cloud-free Sentinel-2 imagery from Google Earth Engine"""

    def __init__(self):
        """Initialize Earth Engine"""
        try:
            ee.Initialize()
            print("✓ Earth Engine initialized successfully")
        except Exception as e:
            print("✗ Earth Engine initialization failed")
            print("Run: earthengine authenticate")
            raise e

    def get_cloud_free_composite(self, roi, start_date, end_date,
                                  cloud_percentage=10, scale=10):
        """
        Get cloud-free Sentinel-2 composite for a region

        Args:
            roi: Earth Engine Geometry (region of interest)
            start_date: Start date string (YYYY-MM-DD)
            end_date: End date string (YYYY-MM-DD)
            cloud_percentage: Maximum cloud percentage (default: 10)
            scale: Resolution in meters (default: 10 for Sentinel-2)

        Returns:
            ee.Image: Cloud-free RGB composite
        """
        print(f"\n  Querying Sentinel-2 images...")
        print(f"  Date range: {start_date} to {end_date}")
        print(f"  Max cloud percentage: {cloud_percentage}%")

        # Load Sentinel-2 Surface Reflectance collection
        collection = ee.ImageCollection('COPERNICUS/S2_SR') \
            .filterBounds(roi) \
            .filterDate(start_date, end_date) \
            .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', cloud_percentage))

        # Count available images
        count = collection.size().getInfo()
        print(f"  Found {count} images matching criteria")

        if count == 0:
            print("  ⚠ Warning: No images found. Try:")
            print("    - Expanding date range")
            print("    - Increasing cloud percentage threshold")
            print("    - Checking if ROI is valid")
            return None

        # Select RGB bands (B4=Red, B3=Green, B2=Blue)
        collection = collection.select(['B4', 'B3', 'B2'])

        # Create median composite (automatically removes clouds)
        composite = collection.median().clip(roi)

        # Scale to 0-255 range for visualization
        composite = composite.visualize(
            min=0,
            max=3000,
            gamma=1.4
        )

        print(f"  ✓ Composite created successfully")
        return composite

    def assess_image_quality(self, image, roi):
        """
        Assess quality of the composite image

        Args:
            image: ee.Image to assess
            roi: Region of interest

        Returns:
            dict: Quality metrics
        """
        # Calculate statistics
        stats = image.reduceRegion(
            reducer=ee.Reducer.mean().combine(
                reducer2=ee.Reducer.stdDev(),
                sharedInputs=True
            ),
            geometry=roi,
            scale=30,
            maxPixels=1e9
        ).getInfo()

        quality = {
            'mean': stats.get('vis-red_mean', 0),
            'std': stats.get('vis-red_stdDev', 0)
        }

        print(f"\n  Quality Assessment:")
        print(f"    Mean pixel value: {quality['mean']:.2f}")
        print(f"    Std deviation: {quality['std']:.2f}")

        # Quality warnings
        if quality['mean'] < 30:
            print("    ⚠ Warning: Very dark image (mean < 30)")
        elif quality['mean'] > 200:
            print("    ⚠ Warning: Very bright image (mean > 200)")
        else:
            print("    ✓ Good brightness range")

        if quality['std'] < 10:
            print("    ⚠ Warning: Low contrast (std < 10)")
        else:
            print("    ✓ Good contrast")

        return quality

    def export_to_drive(self, image, roi, description, folder='RSVQA_Sentinel2',
                       scale=10, crs='EPSG:3857'):
        """
        Export image to Google Drive

        Args:
            image: ee.Image to export
            roi: Region of interest
            description: Export task description
            folder: Google Drive folder name
            scale: Resolution in meters
            crs: Coordinate reference system (EPSG:3857 for Web Mercator)

        Returns:
            ee.batch.Task: Export task
        """
        task = ee.batch.Export.image.toDrive(
            image=image,
            description=description,
            folder=folder,
            scale=scale,
            region=roi,
            crs=crs,
            fileFormat='GeoTIFF',
            maxPixels=1e9,
            formatOptions={
                'cloudOptimized': True
            }
        )

        task.start()
        print(f"\n  ✓ Export task started: {description}")
        print(f"    Task ID: {task.id}")
        print(f"    Check status at: https://code.earthengine.google.com/tasks")

        return task

    def download_to_local(self, image, roi, output_path, scale=10):
        """
        Download image directly to local file

        Args:
            image: ee.Image to download
            roi: Region of interest
            output_path: Local file path
            scale: Resolution in meters
        """
        print(f"\n  Downloading to: {output_path}")

        try:
            # Use geemap for local download
            geemap.ee_export_image(
                image,
                filename=output_path,
                scale=scale,
                region=roi,
                file_per_band=False
            )
            print(f"  ✓ Download complete")
        except Exception as e:
            print(f"  ✗ Download failed: {str(e)}")
            print(f"  Tip: Use export_to_drive() instead for large areas")

    def process_location(self, name, bbox, start_date, end_date,
                        cloud_percentage=10, export_method='drive',
                        output_dir='./downloads'):
        """
        Process a single location

        Args:
            name: Location name
            bbox: Bounding box [lon_min, lat_min, lon_max, lat_max]
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)
            cloud_percentage: Max cloud percentage
            export_method: 'drive' or 'local'
            output_dir: Output directory for local downloads

        Returns:
            dict: Processing results
        """
        print(f"\n{'='*70}")
        print(f"Processing: {name}")
        print(f"{'='*70}")

        # Create ROI
        roi = ee.Geometry.Rectangle(bbox)
        print(f"  Bounding box: {bbox}")

        # Get composite
        composite = self.get_cloud_free_composite(
            roi, start_date, end_date, cloud_percentage
        )

        if composite is None:
            return {
                'name': name,
                'success': False,
                'error': 'No images found'
            }

        # Assess quality
        quality = self.assess_image_quality(composite, roi)

        # Export
        if export_method == 'drive':
            task = self.export_to_drive(
                composite,
                roi,
                description=f'{name}_sentinel2',
                folder='RSVQA_Sentinel2'
            )
            result = {
                'name': name,
                'success': True,
                'task_id': task.id,
                'quality': quality
            }
        else:  # local
            output_path = Path(output_dir) / f'{name}_sentinel2.tif'
            output_path.parent.mkdir(parents=True, exist_ok=True)
            self.download_to_local(composite, roi, str(output_path))
            result = {
                'name': name,
                'success': True,
                'output_path': str(output_path),
                'quality': quality
            }

        return result


def load_locations_from_json(json_path):
    """Load location definitions from JSON file"""
    with open(json_path, 'r') as f:
        config = json.load(f)
    return config


def interactive_mode():
    """Interactive mode to define locations"""
    print("\n" + "="*70)
    print("Interactive Sentinel-2 Download")
    print("="*70)

    locations = []

    while True:
        print("\n" + "-"*70)
        name = input("Location name (or 'done' to finish): ").strip()
        if name.lower() == 'done':
            break

        print("\nEnter bounding box coordinates:")
        lon_min = float(input("  Longitude min: "))
        lat_min = float(input("  Latitude min: "))
        lon_max = float(input("  Longitude max: "))
        lat_max = float(input("  Latitude max: "))
        bbox = [lon_min, lat_min, lon_max, lat_max]

        # Date range
        print("\nDate range (YYYY-MM-DD):")
        start_date = input("  Start date [2023-01-01]: ").strip() or "2023-01-01"
        end_date = input("  End date [2023-12-31]: ").strip() or "2023-12-31"

        # Cloud threshold
        cloud_pct = input("  Max cloud percentage [10]: ").strip()
        cloud_pct = int(cloud_pct) if cloud_pct else 10

        locations.append({
            'name': name,
            'bbox': bbox,
            'start_date': start_date,
            'end_date': end_date,
            'cloud_percentage': cloud_pct
        })

        print(f"\n✓ Added location: {name}")

    return locations


def get_predefined_locations():
    """Get predefined example locations"""
    return [
        {
            'name': 'newyork_manhattan',
            'bbox': [-74.05, 40.7, -73.95, 40.8],
            'start_date': '2023-06-01',
            'end_date': '2023-09-30',
            'cloud_percentage': 10
        },
        {
            'name': 'iowa_farmland',
            'bbox': [-93.65, 41.55, -93.55, 41.65],
            'start_date': '2023-07-01',
            'end_date': '2023-09-30',
            'cloud_percentage': 10
        },
        {
            'name': 'amazon_rainforest',
            'bbox': [-60.1, -3.1, -60.0, -3.0],
            'start_date': '2023-06-01',
            'end_date': '2023-09-30',
            'cloud_percentage': 15
        },
        {
            'name': 'california_coast',
            'bbox': [-122.5, 37.7, -122.4, 37.8],
            'start_date': '2023-06-01',
            'end_date': '2023-09-30',
            'cloud_percentage': 5
        },
        {
            'name': 'arizona_desert',
            'bbox': [-111.1, 33.4, -111.0, 33.5],
            'start_date': '2023-06-01',
            'end_date': '2023-09-30',
            'cloud_percentage': 5
        }
    ]


def main():
    parser = argparse.ArgumentParser(
        description='Download cloud-free Sentinel-2 imagery from Google Earth Engine',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
    # Interactive mode
    python download_sentinel2_gee.py --interactive

    # Use predefined locations
    python download_sentinel2_gee.py --predefined --export_method drive

    # Load from JSON config
    python download_sentinel2_gee.py --config locations.json --export_method local

JSON Config Format:
    {
        "locations": [
            {
                "name": "newyork",
                "bbox": [-74.05, 40.7, -73.95, 40.8],
                "start_date": "2023-06-01",
                "end_date": "2023-09-30",
                "cloud_percentage": 10
            }
        ]
    }

Tips for selecting good locations:
    1. Start with 0.1° × 0.1° areas (about 11km × 11km)
    2. Choose summer months (Jun-Sep) for less cloud cover
    3. Coastal/tropical areas: increase cloud percentage to 15-20
    4. Desert/arid areas: can use lower cloud percentage (5)
    5. Use Google Earth to preview areas first
        """
    )

    parser.add_argument('--interactive', action='store_true',
                       help='Interactive mode to define locations')
    parser.add_argument('--predefined', action='store_true',
                       help='Use predefined example locations')
    parser.add_argument('--config', type=str,
                       help='JSON config file with locations')
    parser.add_argument('--export_method', type=str, default='drive',
                       choices=['drive', 'local'],
                       help='Export to Google Drive or download locally')
    parser.add_argument('--output_dir', type=str, default='./downloads',
                       help='Output directory for local downloads')

    args = parser.parse_args()

    # Get locations
    if args.interactive:
        locations = interactive_mode()
    elif args.predefined:
        locations = get_predefined_locations()
        print("\n" + "="*70)
        print("Using Predefined Locations:")
        print("="*70)
        for loc in locations:
            print(f"  - {loc['name']}: {loc['bbox']}")
    elif args.config:
        config = load_locations_from_json(args.config)
        locations = config['locations']
        print(f"\nLoaded {len(locations)} locations from {args.config}")
    else:
        print("Error: Must specify --interactive, --predefined, or --config")
        return

    if not locations:
        print("No locations to process")
        return

    # Initialize downloader
    print("\nInitializing Earth Engine...")
    downloader = Sentinel2Downloader()

    # Process each location
    results = []
    for location in locations:
        result = downloader.process_location(
            name=location['name'],
            bbox=location['bbox'],
            start_date=location['start_date'],
            end_date=location['end_date'],
            cloud_percentage=location.get('cloud_percentage', 10),
            export_method=args.export_method,
            output_dir=args.output_dir
        )
        results.append(result)

        # Small delay between requests
        time.sleep(2)

    # Summary
    print("\n" + "="*70)
    print("DOWNLOAD SUMMARY")
    print("="*70)

    successful = [r for r in results if r['success']]
    failed = [r for r in results if not r['success']]

    print(f"\nSuccessful: {len(successful)}/{len(results)}")
    for r in successful:
        print(f"  ✓ {r['name']}")
        if args.export_method == 'drive':
            print(f"    Task ID: {r['task_id']}")
        else:
            print(f"    Saved to: {r['output_path']}")
        print(f"    Quality: mean={r['quality']['mean']:.1f}, std={r['quality']['std']:.1f}")

    if failed:
        print(f"\nFailed: {len(failed)}")
        for r in failed:
            print(f"  ✗ {r['name']}: {r['error']}")

    if args.export_method == 'drive':
        print("\n" + "="*70)
        print("NEXT STEPS:")
        print("="*70)
        print("1. Check export status at: https://code.earthengine.google.com/tasks")
        print("2. Wait for tasks to complete (usually 5-30 minutes)")
        print("3. Download from Google Drive folder: RSVQA_Sentinel2")
        print("4. Run preprocessing: python prepare_sentinel2_data.py")
    else:
        print("\n" + "="*70)
        print("NEXT STEPS:")
        print("="*70)
        print(f"1. Check downloads in: {args.output_dir}")
        print("2. Run preprocessing: python prepare_sentinel2_data.py")

    # Save results to JSON
    results_file = 'download_results.json'
    with open(results_file, 'w') as f:
        json.dump({
            'timestamp': datetime.now().isoformat(),
            'export_method': args.export_method,
            'results': results
        }, f, indent=2)
    print(f"\nResults saved to: {results_file}")


if __name__ == '__main__':
    main()
